### ISSUE TITLE
sendMessage helper function

### ISSUE THREAD

**Issue by eringiso** [airgram/airgram#99]
**Issue type:**

[ x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Airgram version:**

[ x] `latest`
[ ] `@next`
[ ] `2.3.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Hi i wrote this function that returns permanent Id of message and the message itself
this function works fine when i call it first time but when i call it the next time
it's returning the temp id of previous message call so it goes to the else statement and 
and doesn't return the permanent id

what am i doing wrong  here?

```
async function sendMessage(chatId, message) {
    return new Promise(async (res, rej) => {
      await airgram.api.sendMessage({

        chatId: chatId,
        replyToMessageId: 0,
        disableNotification: false,
        fromBackground: false,
        inputMessageContent: {
          _: "inputMessageText",
          text: {
            text: message
          }
        }
      }).then((x) => {
        console.log("NEW MESSAGE TEMP ID :");
        console.log(x.response.id);
        console.log("--------------------------here response id works for both first and second calls----------------------------------");
        return x
      }).then(async (x) => {
        const {
          response
        } = x


        //checkin if the message is sent succesfully
        await airgram.on("updateMessageSendSucceeded", async ({
          update
        }) => {

          const {
            message
          } = update

          console.log( response.id + " " + update.oldMessageId);
          console.log("--------------------------here response id comes from the previous call ----------------------------------");
          if (response.id == update.oldMessageId) {
            let m = {
              messageId: message.id,
              message: message.content.text.text
            }
            console.log("message sent with permanent id : " + JSON.stringify(m));
            res(m)
          } else {
            console.log(" ---------------------CHECK HERE BOTH ARE DIFFERENT----------------------->");
            console.log(response.id.toString() + "----" + update.oldMessageId.toString());
          }

        })

      }).catch(console.error)

    })


  }

```

<!--
    To answer those questions you need to put "x" inside the square brackets, for example:
    [x] question
    [ ] bug report

    Also, please format your code properly (by taking code blocks into ```ts .... ```)
!>

**Comment by esindger** [comment]
Hello,
you set middleware every time when you send message. It's not correct behaviour. Middleware must be set only once after creating instance of Airgram.

Try this code:

```ts
interface Deferred<T> {
  promise: Promise<void>
  resolve: (value: T) => unknown
  reject: (error: Error) => unknown
}

function createDeferred<T> (): Deferred<T> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const deferred: Record<string, any> = {}
  deferred.promise = new Promise<T>((resolve, reject) => {
    deferred.resolve = resolve
    deferred.reject = reject
  })
  return deferred as Deferred<T>
}

const airgram = new Airgram({
  // config
})

const pendingMessages = new Map<number, Deferred<Message>>()
airgram.on('sendMessage', async (ctx, next) => {
  const deferred = createDeferred<Message>()
  ctx.setState({
    onSuccess: deferred.promise
  })
  await next()
  const message = toObject(ctx)
  switch (message.sendingState?._) {
    case 'messageSendingStatePending': {
      pendingMessages.set(message.id, deferred)
      break
    }
    case 'messageSendingStateFailed': {
      deferred.reject(new Error('Sending state: messageSendingStateFailed'))
      break
    }
    default: {
      deferred.resolve(message)
    }
  }
})
airgram.on('updateMessageSendSucceeded', async (ctx, next) => {
  const { oldMessageId, message } = ctx.update
  const deferred = pendingMessages.get(oldMessageId)
  if (deferred) {
    deferred.resolve(message)
    pendingMessages.delete(oldMessageId)
  }
  return next()
})

const sendMessage = async (chatId, message) => {
  return airgram.api.sendMessage({
    // options
  }).then((ctx) => {
    const { getState } = ctx
    const { onSuccess } = getState()
    return onSuccess
  }).catch(console.error)
}
```

**Comment by eringiso** [comment]
Hi,
Sorry to bother you but I have never used typescript before it seems i've opened the issue in the wrong repo because i'm using the plain javascript example of airgram  even then  i went ahead and tried to implement the above code with plain js but i get error ctx.setState is not a function on sendMessage 
i've read the doc on sendMessage for both tdlib and airgram and the response has a setState function only in airgram

so i think i'm getting the error coz i'm not using the typescript version of airgram if not
could you please tell me what i'm doin wrong or give an example of how i could do this in plain js

here is what i'm doing since there is no interface types or deferred object in plain js i just created
a function createDeferred just like you did above

```
function createDeferred(){
  const deferred = {}
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve
    deferred.reject = reject
  })
  return deferred
}

const pendingMessages = new Map()

airgram.on('sendMessage', async (ctx, next) => {
  const deferred = createDeferred()
  ctx.setState({    //----------getting error here--------------//
    onSuccess: deferred.promise
  })
  await next()
  const message = toObject(ctx)
  switch (message.sendingState) {
    case 'messageSendingStatePending': {
      pendingMessages.set(message.id, deferred)
      break
    }
    case 'messageSendingStateFailed': {
      deferred.reject(new Error('Sending state: messageSendingStateFailed'))
      break
    }
    default: {
      deferred.resolve(message)
    }
  }
})

```

**Comment by eringiso** [comment]
Hi,
I've resolved it i shouldn't be setting the state to the ctx directly
instead to the state={} like this

```
airgram.on('sendMessage', async (ctx, next) => {
  const deferred = createDeferred()
  const { state } = ctx
  state.setState({onSuccess : deferred.promise})
  await next()
  const message = toObject(ctx)
  switch (message.sendingState) {
    case 'messageSendingStatePending': {
      pendingMessages.set(message.id, deferred)
      break
    }
    case 'messageSendingStateFailed': {
      deferred.reject(new Error('Sending state: messageSendingStateFailed'))
      break
    }
    default: {
      deferred.resolve(message)
    }
  }
})

```

**Comment by eringiso** [comment]
Hi before i close this issue i want to ask this

i'm trying to check messages from 3 different chats
so i'm checking their chatid's and then printing to log

but everytime i check more than one id i get updates from all the chats 
instead of the chats i specified

```
airgram.on('updateNewMessage', async ({
  update
}) => {

  const { message } = update

  if (message.chatId === x) {

    console.log('[new message]', message.content.text.text)
   // here  i get message updates only from x channel
  }

  if (message.chatId === x || y || z) {

    console.log('[new message]', message.content.text.text)
   // here  i get message updates  from all the  channels and groups instead of x,y,z

  }


})
```

the problem is if i use or condition i'm getting updates from all the channels
is there anything i'm doing wrong here

**Comment by esindger** [comment]
Here is JavaScript version of my example.
```js
function createDeferred() {
  const deferred = {}
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve
    deferred.reject = reject
  })
  return deferred
}

const airgram = new Airgram({
  // config
})

const pendingMessages = new Map()

airgram.on('sendMessage', async (ctx, next) => {
  const deferred = createDeferred()
  ctx.setState({
    onSuccess: deferred.promise
  })
  await next()
  const message = toObject(ctx)
  switch (message.sendingState && message.sendingState._) {
    case 'messageSendingStatePending': {
      pendingMessages.set(message.id, deferred)
      break
    }
    case 'messageSendingStateFailed': {
      deferred.reject(new Error('Sending state: messageSendingStateFailed'))
      break
    }
    default: {
      deferred.resolve(message)
    }
  }
})

airgram.on('updateMessageSendSucceeded', async (ctx, next) => {
  const { oldMessageId, message } = ctx.update
  const deferred = pendingMessages.get(oldMessageId)
  if (deferred) {
    deferred.resolve(message)
    pendingMessages.delete(oldMessageId)
  }
  return next()
})

const sendMessage = (chatId, message) => {
  return airgram.api.sendMessage({
    // options
  }).then((ctx) => {
    const { getState } = ctx
    const { onSuccess } = getState()
    return onSuccess
  })
}
```

**Comment by esindger** [comment]
Please look at this part of your code:
```js
if (message.chatId === x || y || z) {
    console.log('[new message]', message.content.text.text)
   // here  i get message updates  from all the  channels and groups instead of x,y,z
  }
```
It's the same as:
```js
if ((message.chatId === x) || (y != null) || (z != null)) {
    // ...
  }
```

The `message.chatId === x || y || z` condition will return `true`, if `y`  or `z` have defined.

**Comment by eringiso** [comment]
Hi thanks a lot now i know what i'm doing wrong

also since you gave a plain js example, that is exactly what i did before
but i get error when set state to the ctx 

```
  ctx.setState({    
    onSuccess: deferred.promise
  })
```

so this is what i did, please take a look here
https://github.com/airgram/airgram/issues/99#issuecomment-629111456
and it works exactly what i wanted it to do i just want to know if what
i did is the right thing

**Comment by esindger** [comment]
> ```
>   const { state } = ctx
>   state.setState({onSuccess : deferred.promise})
> ```

Context does not have `state` property. You should use `ctx.getState()` and `ctx.setState()`. Here is [documentation](https://airgram.io/guides/basic-concepts/operation-state).

**Comment by eringiso** [comment]
I tried setting the state to the ctx 
```
  ctx.setState({
    onSuccess: deferred.promise
  })
```
but i get error 
TypeError: ctx.setState is not a function

**Comment by esindger** [comment]
There was old version of Airgram in dependencies.

Please fix your package.json:
```
"dependencies": {
    "airgram": "^3.0.2"
  },
```

**Comment by eringiso** [comment]
Thanks a lot
it works perfectly fine now

**Comment by jumoog** [comment]
thank you for your code snippet! i have used await before. But now I really wait until the message is really sent.
